src/backend/codegen/CMakeLists.txt

```diff
@@ -162,26 +183,121 @@ target_link_libraries(
+add_simple_gtest(instance_method_wrappers_unittest tests/instance_method_wrappers_unittest.cc)
+
+# Usage add_cmock_gtest ${TEST_NAME} ${TEST_SOURCES} ${MOCK_DIR}/hello_mock.o ${MOCK_DIR}/world_mock.o)
+function(add_cmockery_gtest TEST_NAME TEST_SOURCES)
+    set(FILES_TO_LINK ${OBJFILES})
+
+    foreach(MOCK_OBJ_NAME ${ARGN})
+        string(REPLACE "_mock" "" REAL_OBJ_NAME ${MOCK_OBJ_NAME})
+        string(REPLACE "${MOCK_DIR}" "${TOP_SRC_DIR}/src" REAL_OBJ_NAME ${REAL_OBJ_NAME})
+        list(REMOVE_ITEM FILES_TO_LINK ${REAL_OBJ_NAME})
+
+        if(EXISTS ${MOCK_OBJ_NAME})
+            list(APPEND FILES_TO_LINK ${MOCK_OBJ_NAME})
+        else()
+            string(REPLACE ".o" ".c" REAL_SRC_NAME ${REAL_OBJ_NAME})
+            string(REPLACE ".o" ".c" MOCK_SRC_NAME ${MOCK_OBJ_NAME})
+            get_filename_component(REAL_SRC_ABS_PATH ${REAL_SRC_NAME} ABSOLUTE)
+            execute_process(COMMAND python mocker.py ${REAL_SRC_ABS_PATH}
+                WORKING_DIRECTORY ${MOCK_DIR})
+            list(APPEND FILES_TO_LINK ${MOCK_SRC_NAME})
+        endif()
+    endforeach()
+
+    add_executable(${TEST_NAME} EXCLUDE_FROM_ALL 
+        ${TEST_SOURCES}
+        ${GPCODEGEN_SRC}
+        ${FILES_TO_LINK} 
+        ${MOCK_OBJS} 
+        ${CMOCKERY_OBJS}
+    )
+    # Bring these from $ENV{LIBS}
+    target_link_libraries(${TEST_NAME} "-ldl -lbz2 -lxml2 -lz -lm -lcurl -L../../port -lpgport_srv" gpcodegen gtest)
     add_test(${TEST_NAME} ${TEST_NAME})
-    add_dependencies(unittest-check ${TEST_NAME})
-endforeach()
+    add_dependencies(check ${TEST_NAME})
+endfunction(add_cmockery_gtest)
+
+# Get the list for all real objects from objfiles.txt for CMockery integration
+set(TXT_OBJFILE ${TOP_SRC_DIR}/src/backend/objfiles.txt)
+if(EXISTS ${TXT_OBJFILE})
+    file(GLOB_RECURSE GPOPT_OBJS RELATIVE ${TOP_SRC_DIR} ${TOP_SRC_DIR}/src/backend/gpopt/*.o)
+    set(EXCL_OBJS
+        src/backend/main/main.o 
+        ${GPOPT_OBJS})
+
+    file(READ ${TXT_OBJFILE} OBJFILES)
+    string(REPLACE "\n" ";" OBJFILES "${OBJFILES}")
+    string(REPLACE " " ";" OBJFILES "${OBJFILES}")
+    set(OBJFILES ${OBJFILES})
+
+    foreach(F ${OBJFILES})
+        get_filename_component(_F ${F} NAME)
+        if ("${_F}" STREQUAL "objfiles.txt")
+            list(REMOVE_ITEM OBJFILES ${F})
+        endif()
+    endforeach()
+
+    foreach(OBJ ${EXCL_OBJS})
+        list(REMOVE_ITEM OBJFILES ${OBJ})
+    endforeach()
+    prepend_path(OBJFILES ${TOP_SRC_DIR} ${OBJFILES})
+
+
+    set(MOCK_OBJS "")
+    foreach(OBJ
+        ${TOP_SRC_DIR}/src/test/unit/mock/main_mock.o
+        ${TOP_SRC_DIR}/src/test/unit/mock/gpopt_mock.o 
+    )
+        if(EXISTS ${OBJ})
+            set(MOCK_OBJS ${OBJ} ${MOCK_OBJS})
+        endif()
+    endforeach()
+
+    set(CMOCKERY_DIR ${TOP_SRC_DIR}/src/test/unit/cmockery)
+    set(CMOCKERY_OBJS ${CMOCKERY_DIR}/cmockery.o)
+endif()
+
+# Add CMockery tests
+if(EXISTS ${TXT_OBJFILE})
+    add_cmockery_gtest(codegen_framework_unittest 
```

Comment:
We should rename this to codegen_framework_unittest.t so that git ignore rules apply and these files don't clutter the git status